addons/mail/models/fetchmail.py

```diff
@@ -190,102 +238,101 @@ def button_confirm_login(self):
+            server_type_and_name = server.server_type, server.name  # avoid reading this after each commit
+            _logger.info('Start checking for new emails on %s server %s', *server_type_and_name)
             count, failed = 0, 0
-            imap_server = None
-            pop_server = None
-            connection_type = server._get_connection_type()
-            if connection_type == 'imap':
-                try:
-                    imap_server = server.connect()
-                    imap_server.select()
-                    result, data = imap_server.search(None, '(UNSEEN)')
-                    for num in data[0].split():
-                        res_id = None
-                        result, data = imap_server.fetch(num, '(RFC822)')
-                        imap_server.store(num, '-FLAGS', '\\Seen')
-                        try:
-                            res_id = MailThread.with_context(**additionnal_context).message_process(server.object_id.model, data[0][1], save_original=server.original, strip_attachments=(not server.attach))
-                        except Exception:
-                            _logger.info('Failed to process mail from %s server %s.', server.server_type, server.name, exc_info=True)
-                            failed += 1
-                        imap_server.store(num, '+FLAGS', '\\Seen')
-                        self._cr.commit()
-                        count += 1
-                    _logger.info("Fetched %d email(s) on %s server %s; %d succeeded, %d failed.", count, server.server_type, server.name, (count - failed), failed)
-                except Exception as e:
-                    if raise_exception:
-                        raise ValidationError(_("Couldn't get your emails. Check out the error message below for more info:\n%s", e)) from e
+
+            # processing messages in a separate transaction to keep lock on the server
+            server_connection = None
+            message_cr = None
+            try:
+                server_connection = server.connect()
+                message_cr = self.env.registry.cursor()
+                MailThread = server.env['mail.thread'].with_env(self.env(cr=message_cr)).with_context(default_fetchmail_server_id=server.id)
+                thread_process_message = functools.partial(
+                    MailThread.message_process,
+                    model=server.object_id.model,
+                    save_original=server.original,
+                    strip_attachments=(not server.attach),
+                )
+                unread_message_count = server_connection.check_unread_messages()
+                _logger.debug('%d unread messages on %s server %s.', unread_message_count, *server_type_and_name)
+                total_remaining += unread_message_count
+                for message_num, message in server_connection.retrieve_unread_messages():
+                    _logger.debug('Fetched message %r on %s server %s.', message_num, *server_type_and_name)
```

Comment:
Do we need to debug all messages ?

Reply:
It's in the debug level, so only logged when debugging.